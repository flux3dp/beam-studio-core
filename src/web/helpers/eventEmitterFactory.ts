import EventEmitter from 'eventemitter3';

const eventEmitters = {
  'flux-id': null,
  'top-bar': null,
  'top-bar-hints': null,
  'time-estimation-button': null,
};
export default {
  createEventEmitter: (type?: string): EventEmitter => {
    if (!type) return new EventEmitter();
    if (!eventEmitters[type]) eventEmitters[type] = new EventEmitter();
    return eventEmitters[type];
  },
};
