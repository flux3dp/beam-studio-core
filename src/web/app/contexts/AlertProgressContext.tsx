import React, { createContext } from 'react';

import AlertConstants from 'app/constants/alert-constants';
import eventEmitterFactory from 'helpers/eventEmitterFactory';
import i18n from 'helpers/i18n';
import ProgressConstants from 'app/constants/progress-constants';
import { IAlert } from 'interfaces/IAlert';
import { IProgressDialog } from 'interfaces/IProgress';

const LANG = i18n.lang.alert;
let progressID = 0;

interface IAlertProgressContext {
  alertProgressStack: (IAlert | IProgressDialog)[];
  popFromStack: () => void;
  popById: (id: string) => void;
}

export const AlertProgressContext = createContext<IAlertProgressContext>({
  alertProgressStack: [],
  popFromStack: () => {},
  popById: () => {},
});

export const eventEmitter = eventEmitterFactory.createEventEmitter();

interface State {
  alertProgressStack: (IAlert | IProgressDialog)[],
}

export class AlertProgressContextProvider extends React.Component<unknown, State> {
  constructor(props: unknown) {
    super(props);
    this.state = {
      alertProgressStack: [],
    };

    eventEmitter.on('OPEN_PROGRESS', this.openProgress.bind(this));
    eventEmitter.on('POP_LAST_PROGRESS', this.popLastProgress.bind(this));
    eventEmitter.on('UPDATE_PROGRESS', this.updateProgress.bind(this));
    eventEmitter.on('POP_BY_ID', this.popById.bind(this));
    eventEmitter.on('POP_UP', this.popUp.bind(this));
    eventEmitter.on('CHECK_ID_EXIST', this.checkIdExist.bind(this));
  }

  componentWillUnmount() {
    eventEmitter.removeAllListeners();
  }

  popFromStack = (): void => {
    const { alertProgressStack } = this.state;
    alertProgressStack.pop();
    this.forceUpdate();
  };

  popById = (id: string): void => {
    const { alertProgressStack } = this.state;
    this.setState({
      alertProgressStack: alertProgressStack.filter((item) => item.id !== id),
    });
  };

  checkIdExist = (id: string, response: {
    idExist: boolean,
  }): void => {
    const { alertProgressStack } = this.state;
    const res = alertProgressStack.filter((item) => {
      const { id: itemId } = item;
      return itemId === id && !('isProgress' in item);
    });
    response.idExist = res.length > 0;
  };

  pushToStack = (item: (IAlert | IProgressDialog)): void => {
    if (item.id) {
      // eslint-disable-next-line no-console
      console.log('alert/progress poped', item.id);
    }
    const { alertProgressStack } = this.state;
    alertProgressStack.push(item);
    this.forceUpdate();
  };

  openProgress = (args: IProgressDialog): void => {
    const { type } = args;
    let { id, message, caption } = args;
    if (!id) {
      id = `progress_${progressID}`;
      progressID = (progressID + 1) % 100000;
    }
    message = message || '';
    caption = caption || '';

    this.pushToStack({
      ...args,
      id,
      type,
      caption,
      message,
      isProgress: true,
    });
  };

  popLastProgress = (): void => {
    const { alertProgressStack } = this.state;
    let i;
    for (i = alertProgressStack.length - 1; i >= 0; i -= 1) {
      if ('isProgress' in alertProgressStack[i]) {
        break;
      }
    }
    if (i >= 0) {
      alertProgressStack.splice(i, 1);
      this.setState({ alertProgressStack });
    }
  };

  updateProgress = (id: string, args: IProgressDialog): void => {
    const { alertProgressStack } = this.state;
    const targetObjects = alertProgressStack.filter((item) => {
      const { id: itemId } = item;
      return 'isProgress' in item && itemId === id;
    });
    if (targetObjects.length === 0) {
      return;
    }
    const targetObject = targetObjects[targetObjects.length - 1];
    if (targetObject.type === ProgressConstants.NONSTOP && !args.caption && args.message) {
      // eslint-disable-next-line no-param-reassign
      args.caption = args.message;
    }
    Object.assign(targetObject, args);
    this.setState({ alertProgressStack });
  };

  popUp = (args: IAlert): void => {
    const { type } = args;
    let { message, caption } = args;
    message = message || '';
    switch (type) {
      case AlertConstants.SHOW_POPUP_INFO:
        caption = caption || LANG.info;
        break;
      case AlertConstants.SHOW_POPUP_WARNING:
        caption = caption || LANG.warning;
        break;
      case AlertConstants.SHOW_POPUP_ERROR:
        caption = caption || LANG.error;
        break;
      default:
        break;
    }
    const { buttons, checkbox } = this.buttonsGenerator(args);
    const checkboxText = checkbox ? checkbox.text : null;
    const checkboxCallbacks = checkbox ? checkbox.callbacks : null;

    this.pushToStack({
      ...args,
      caption,
      message,
      buttons,
      checkboxText,
      checkboxCallbacks,
    });
  };

  buttonsGenerator = (args: IAlert): { buttons, checkbox } => {
    const { checkbox } = args;
    let { buttons } = args;
    if (buttons) {
      return { buttons, checkbox };
    }
    const {
      id,
      buttonType,
    } = args;
    let {
      buttonLabels,
      callbacks,
      primaryButtonIndex,
      onYes,
      onNo,
      onConfirm,
      onRetry,
      onCancel,
    } = args;
    switch (buttonType) {
      case AlertConstants.YES_NO:
        onYes = onYes || (() => { });
        onNo = onNo || (() => { });
        buttonLabels = [LANG.yes, LANG.no];
        callbacks = [onYes, onNo];
        primaryButtonIndex = primaryButtonIndex || 0;
        if (checkbox) {
          const onCheckedYes = checkbox.onYes ? checkbox.onYes : () => { };
          const onCheckedNo = checkbox.onNo ? checkbox.onNo : () => { };
          checkbox.callbacks = [onCheckedYes, onCheckedNo];
        }
        break;
      case AlertConstants.YES_NO_CUSTOM:
        onYes = onYes || (() => { });
        onNo = onNo || (() => { });
        primaryButtonIndex = primaryButtonIndex || 0;
        if (!buttonLabels) {
          buttonLabels = [LANG.yes, LANG.no];
          callbacks = [onYes, onNo];
          if (checkbox) {
            const onCheckedYes = checkbox.onYes ? checkbox.onYes : () => { };
            const onCheckedNo = checkbox.onNo ? checkbox.onNo : () => { };
            checkbox.callbacks = [onCheckedYes, onCheckedNo];
          }
        } else if (typeof buttonLabels === 'string') {
          buttonLabels = [LANG.yes, LANG.no, buttonLabels];
          callbacks = callbacks || (() => { });
          callbacks = [onYes, onNo, (callbacks as () => void)];
        } else {
          buttonLabels = [LANG.yes, LANG.no, ...buttonLabels];
          callbacks = [onYes, onNo, ...(callbacks as (() => void)[])];
        }
        break;
      case AlertConstants.CONFIRM_CANCEL:
        onConfirm = onConfirm || (() => { });
        onCancel = onCancel || (() => { });
        buttonLabels = [LANG.confirm, LANG.cancel];
        primaryButtonIndex = primaryButtonIndex || 0;
        callbacks = [onConfirm, onCancel];
        if (checkbox) {
          const onCheckedConfirm = checkbox.onConfirm ? checkbox.onConfirm : () => { };
          const onCheckedCancel = checkbox.onCancel ? checkbox.onCancel : () => { };
          checkbox.callbacks = [onCheckedConfirm, onCheckedCancel];
        }
        break;
      case AlertConstants.RETRY_CANCEL:
        onRetry = onRetry || (() => { });
        onCancel = onCancel || (() => { });
        buttonLabels = [LANG.retry, LANG.cancel];
        primaryButtonIndex = primaryButtonIndex || 0;
        callbacks = [onRetry, onCancel];
        if (checkbox) {
          const onCheckedRetry = checkbox.onRetry ? checkbox.onRetry : () => { };
          const onCheckedCancel = checkbox.onCancel ? checkbox.onCancel : () => { };
          checkbox.callbacks = [onCheckedRetry, onCheckedCancel];
        }
        break;
      case AlertConstants.CUSTOM_CANCEL:
        onCancel = onCancel || (() => { });
        primaryButtonIndex = primaryButtonIndex || 0;
        if (!buttonLabels) {
          buttonLabels = [LANG.cancel];
          callbacks = [onCancel];
          if (checkbox) {
            const onCheckedCancel = checkbox.onCancel ? checkbox.onCancel : () => { };
            checkbox.callbacks = [onCheckedCancel];
          }
        } else if (typeof buttonLabels === 'string') {
          buttonLabels = [buttonLabels, LANG.cancel];
          callbacks = callbacks || (() => { });
          callbacks = [callbacks as () => void, onCancel];
        } else {
          buttonLabels = [...buttonLabels, LANG.cancel];
          callbacks = [...(callbacks as (() => void)[]), onCancel];
        }
        break;
      default:
        if (!buttonLabels) {
          buttonLabels = [LANG.ok];
          callbacks = callbacks || (() => { });
          if (checkbox) {
            if (!checkbox.callbacks) {
              checkbox.callbacks = [() => { }];
            } else if (typeof checkbox.callbacks === 'function') {
              checkbox.callbacks = [checkbox.callbacks];
            }
          }
        } else if (typeof buttonLabels === 'string') {
          buttonLabels = [buttonLabels];
          callbacks = callbacks || (() => { });
          if (checkbox) {
            if (!checkbox.callbacks) {
              checkbox.callbacks = [() => { }];
            } else if (typeof checkbox.callbacks === 'function') {
              checkbox.callbacks = [checkbox.callbacks];
            }
          }
        }
        break;
    }
    buttons = buttonLabels.map((label, i) => {
      const b = {
        label,
        className: (buttonLabels.length === 1 || i === primaryButtonIndex || primaryButtonIndex === undefined) ? 'btn-default primary' : 'btn-default',
        onClick: () => { },
      };
      if (callbacks && typeof callbacks === 'function') {
        b.onClick = () => {
          (callbacks as (id: string) => void)(id);
        };
      } else if (callbacks && callbacks.length > i) {
        b.onClick = () => {
          callbacks[i](id);
        };
      } else if (!callbacks) {
        b.onClick = () => { };
      }
      return b;
    });

    return { buttons, checkbox };
  };

  render(): JSX.Element {
    const { children } = this.props;
    const { alertProgressStack } = this.state;
    return (
      <AlertProgressContext.Provider value={
          {
            alertProgressStack,
            popFromStack: this.popFromStack,
            popById: this.popById,
          }
        }
      >
        {children}
      </AlertProgressContext.Provider>
    );
  }
}
