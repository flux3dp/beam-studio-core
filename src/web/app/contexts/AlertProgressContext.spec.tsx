/* eslint-disable import/first */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

jest.mock('helpers/i18n', () => ({
  lang: {
    alert: {
      info: 'INFO',
      warning: 'WARNING',
      error: 'UH-OH',
      retry: 'Retry',
      confirm: 'Confirm',
      cancel: 'Cancel',
      ok: 'OK',
      yes: 'Yes',
      no: 'No',
    },
  },
}));

import AlertConstants from 'app/constants/alert-constants';
import eventEmitterFactory from 'helpers/eventEmitterFactory';
import ProgressConstants from 'app/constants/progress-constants';
import { AlertProgressContextProvider, AlertProgressContext } from './AlertProgressContext';

const eventEmitter = eventEmitterFactory.createEventEmitter('alert-progress');
const Children = () => {
  const context = React.useContext(AlertProgressContext);
  return (
    <>
      {context.alertProgressStack}
    </>
  );
};

test('should render correctly', () => {
  const wrapper = shallow(
    <AlertProgressContextProvider>
      <Children />
    </AlertProgressContextProvider>,
  );
  expect(toJson(wrapper)).toMatchSnapshot();
  expect(eventEmitter.eventNames().length).toBe(7);

  eventEmitter.emit('OPEN_PROGRESS', {
    id: 'check-status',
    type: ProgressConstants.NONSTOP,
    caption: 'preparing',
    isProgress: true,
  });
  eventEmitter.emit('OPEN_PROGRESS', {
    id: 'get_log',
    type: ProgressConstants.STEPPING,
    message: 'downloading',
    isProgress: true,
    percentage: 0,
  });
  expect(toJson(wrapper)).toMatchSnapshot();

  eventEmitter.emit('UPDATE_PROGRESS', 'get_log1', {
    percentage: 100,
  });
  expect(toJson(wrapper)).toMatchSnapshot();

  eventEmitter.emit('UPDATE_PROGRESS', 'get_log', {
    percentage: 100,
  });
  expect(toJson(wrapper)).toMatchSnapshot();

  eventEmitter.emit('UPDATE_PROGRESS', 'check-status', {
    message: 'prepared',
  });
  expect(toJson(wrapper)).toMatchSnapshot();

  // type: 'SHOW_POPUP_INFO' + button type: 'YES_NO'
  eventEmitter.emit('POP_UP', {
    type: AlertConstants.SHOW_POPUP_INFO,
    message: 'File already exists, do you want to replace it?',
    buttonType: AlertConstants.YES_NO,
    onYes: jest.fn(),
  });
  // type: 'SHOW_POPUP_WARNING' + default button type + NO button labels
  eventEmitter.emit('POP_UP', {
    type: AlertConstants.SHOW_POPUP_WARNING,
    message: 'Some texts were changed to other Fonts when parsing texts to paths and some character may not converted normally.',
    callbacks: jest.fn(),
    checkbox: {
      text: 'Don\'t Show this next time.',
      callbacks: jest.fn(),
    },
  });
  // type: 'SHOW_POPUP_ERROR' + button type: 'RETRY_CANCEL'
  eventEmitter.emit('POP_UP', {
    id: 'monitor-reconnect',
    type: AlertConstants.SHOW_POPUP_ERROR,
    buttonType: AlertConstants.RETRY_CANCEL,
    message: 'The connection with the machine has broken. Do you want to reconnect?',
    onRetry: jest.fn(),
  });
  // default type + button type: 'CONFIRM_CANCEL'
  eventEmitter.emit('POP_UP', {
    buttonType: AlertConstants.CONFIRM_CANCEL,
    message: 'This will load arrangement of presets and replacing customized parameters set in the file, are you sure to proceed?',
    onConfirm: jest.fn(),
  });
  // default type + button type: 'CUSTOM_CANCEL'
  eventEmitter.emit('POP_UP', {
    id: 'latest-firmware',
    message: 'You have the latest Machine firmware',
    caption: 'Machine firmware Update',
    buttonType: AlertConstants.CUSTOM_CANCEL,
    buttonLabels: ['UPDATE'],
    callbacks: jest.fn(),
    onCancel: jest.fn(),
  });
  // has buttons
  eventEmitter.emit('POP_UP', {
    message: 'Unable to find machine ',
    buttons: [{
      label: 'Set Connection',
      className: 'btn-default primary',
      onClick: jest.fn(),
    },
    {
      label: 'Retry',
      className: 'btn-default primary',
      onClick: jest.fn(),
    }],
  });
  // type: 'SHOW_POPUP_INFO' + default button type + has button labels
  eventEmitter.emit('POP_UP', {
    id: 'machine-info',
    type: AlertConstants.SHOW_POPUP_INFO,
    caption: 'abcde',
    message: 'abcde: 111.222.333.444',
    buttonLabels: ['Test Network', 'OK'],
    callbacks: [
      jest.fn(),
      jest.fn(),
    ],
    primaryButtonIndex: 1,
  });
  expect(toJson(wrapper)).toMatchSnapshot();

  eventEmitter.emit('POP_BY_ID', 'machine-info');
  expect(toJson(wrapper)).toMatchSnapshot();

  const response = {
    idExist: false,
  };
  eventEmitter.emit('CHECK_ID_EXIST', 'latest-firmware', response);
  expect(response.idExist).toBeTruthy();
  eventEmitter.emit('CHECK_ID_EXIST', 'check-status', response);
  expect(response.idExist).toBeFalsy();

  eventEmitter.emit('POP_LAST_PROGRESS');
  expect(toJson(wrapper)).toMatchSnapshot();

  wrapper.unmount();
  expect(eventEmitter.eventNames().length).toBe(0);
});
