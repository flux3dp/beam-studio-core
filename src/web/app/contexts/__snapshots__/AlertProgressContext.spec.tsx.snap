// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<ContextProvider
  value={
    Object {
      "alertProgressStack": Array [],
      "popById": [Function],
      "popFromStack": [Function],
    }
  }
>
  <Children />
</ContextProvider>
`;

exports[`should render correctly 2`] = `
<ContextProvider
  value={
    Object {
      "alertProgressStack": Array [
        Object {
          "caption": "preparing",
          "id": "check-status",
          "isProgress": true,
          "message": "",
          "type": "NONSTOP",
        },
        Object {
          "caption": "",
          "id": "get_log",
          "isProgress": true,
          "message": "downloading",
          "percentage": 0,
          "type": "STEPPING",
        },
      ],
      "popById": [Function],
      "popFromStack": [Function],
    }
  }
>
  <Children />
</ContextProvider>
`;

exports[`should render correctly 3`] = `
<ContextProvider
  value={
    Object {
      "alertProgressStack": Array [
        Object {
          "caption": "preparing",
          "id": "check-status",
          "isProgress": true,
          "message": "",
          "type": "NONSTOP",
        },
        Object {
          "caption": "",
          "id": "get_log",
          "isProgress": true,
          "message": "downloading",
          "percentage": 0,
          "type": "STEPPING",
        },
      ],
      "popById": [Function],
      "popFromStack": [Function],
    }
  }
>
  <Children />
</ContextProvider>
`;

exports[`should render correctly 4`] = `
<ContextProvider
  value={
    Object {
      "alertProgressStack": Array [
        Object {
          "caption": "preparing",
          "id": "check-status",
          "isProgress": true,
          "message": "",
          "type": "NONSTOP",
        },
        Object {
          "caption": "",
          "id": "get_log",
          "isProgress": true,
          "message": "downloading",
          "percentage": 100,
          "type": "STEPPING",
        },
      ],
      "popById": [Function],
      "popFromStack": [Function],
    }
  }
>
  <Children />
</ContextProvider>
`;

exports[`should render correctly 5`] = `
<ContextProvider
  value={
    Object {
      "alertProgressStack": Array [
        Object {
          "caption": "prepared",
          "id": "check-status",
          "isProgress": true,
          "message": "prepared",
          "type": "NONSTOP",
        },
        Object {
          "caption": "",
          "id": "get_log",
          "isProgress": true,
          "message": "downloading",
          "percentage": 100,
          "type": "STEPPING",
        },
      ],
      "popById": [Function],
      "popFromStack": [Function],
    }
  }
>
  <Children />
</ContextProvider>
`;

exports[`should render correctly 6`] = `
<ContextProvider
  value={
    Object {
      "alertProgressStack": Array [
        Object {
          "caption": "prepared",
          "id": "check-status",
          "isProgress": true,
          "message": "prepared",
          "type": "NONSTOP",
        },
        Object {
          "caption": "",
          "id": "get_log",
          "isProgress": true,
          "message": "downloading",
          "percentage": 100,
          "type": "STEPPING",
        },
        Object {
          "buttonType": "YES_NO",
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "Yes",
              "onClick": [Function],
            },
            Object {
              "className": "btn-default",
              "label": "No",
              "onClick": [Function],
            },
          ],
          "caption": "INFO",
          "checkboxCallbacks": null,
          "checkboxText": null,
          "message": "File already exists, do you want to replace it?",
          "onYes": [MockFunction],
          "type": "SHOW_POPUP_INFO",
        },
        Object {
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "OK",
              "onClick": [Function],
            },
          ],
          "callbacks": [MockFunction],
          "caption": "WARNING",
          "checkbox": Object {
            "callbacks": [MockFunction],
            "text": "Don't Show this next time.",
          },
          "checkboxCallbacks": [MockFunction],
          "checkboxText": "Don't Show this next time.",
          "message": "Some texts were changed to other Fonts when parsing texts to paths and some character may not converted normally.",
          "type": "SHOW_POPUP_WARNING",
        },
        Object {
          "buttonType": "RETRY_CANCEL",
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "Retry",
              "onClick": [Function],
            },
            Object {
              "className": "btn-default",
              "label": "Cancel",
              "onClick": [Function],
            },
          ],
          "caption": "UH-OH",
          "checkboxCallbacks": null,
          "checkboxText": null,
          "id": "monitor-reconnect",
          "message": "The connection with the machine has broken. Do you want to reconnect?",
          "onRetry": [MockFunction],
          "type": "SHOW_POPUP_ERROR",
        },
        Object {
          "buttonType": "CONFIRM_CANCEL",
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "Confirm",
              "onClick": [Function],
            },
            Object {
              "className": "btn-default",
              "label": "Cancel",
              "onClick": [Function],
            },
          ],
          "caption": undefined,
          "checkboxCallbacks": null,
          "checkboxText": null,
          "message": "This will load arrangement of presets and replacing customized parameters set in the file, are you sure to proceed?",
          "onConfirm": [MockFunction],
        },
        Object {
          "buttonLabels": Array [
            "UPDATE",
          ],
          "buttonType": "CUSTOM_CANCEL",
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "UPDATE",
              "onClick": [Function],
            },
            Object {
              "className": "btn-default",
              "label": "Cancel",
              "onClick": [Function],
            },
          ],
          "callbacks": [MockFunction],
          "caption": "Machine firmware Update",
          "checkboxCallbacks": null,
          "checkboxText": null,
          "id": "latest-firmware",
          "message": "You have the latest Machine firmware",
          "onCancel": [MockFunction],
        },
        Object {
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "Set Connection",
              "onClick": [MockFunction],
            },
            Object {
              "className": "btn-default primary",
              "label": "Retry",
              "onClick": [MockFunction],
            },
          ],
          "caption": undefined,
          "checkboxCallbacks": null,
          "checkboxText": null,
          "message": "Unable to find machine ",
        },
        Object {
          "buttonLabels": Array [
            "Test Network",
            "OK",
          ],
          "buttons": Array [
            Object {
              "className": "btn-default",
              "label": "Test Network",
              "onClick": [Function],
            },
            Object {
              "className": "btn-default primary",
              "label": "OK",
              "onClick": [Function],
            },
          ],
          "callbacks": Array [
            [MockFunction],
            [MockFunction],
          ],
          "caption": "abcde",
          "checkboxCallbacks": null,
          "checkboxText": null,
          "id": "machine-info",
          "message": "abcde: 111.222.333.444",
          "primaryButtonIndex": 1,
          "type": "SHOW_POPUP_INFO",
        },
      ],
      "popById": [Function],
      "popFromStack": [Function],
    }
  }
>
  <Children />
</ContextProvider>
`;

exports[`should render correctly 7`] = `
<ContextProvider
  value={
    Object {
      "alertProgressStack": Array [
        Object {
          "caption": "prepared",
          "id": "check-status",
          "isProgress": true,
          "message": "prepared",
          "type": "NONSTOP",
        },
        Object {
          "caption": "",
          "id": "get_log",
          "isProgress": true,
          "message": "downloading",
          "percentage": 100,
          "type": "STEPPING",
        },
        Object {
          "buttonType": "YES_NO",
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "Yes",
              "onClick": [Function],
            },
            Object {
              "className": "btn-default",
              "label": "No",
              "onClick": [Function],
            },
          ],
          "caption": "INFO",
          "checkboxCallbacks": null,
          "checkboxText": null,
          "message": "File already exists, do you want to replace it?",
          "onYes": [MockFunction],
          "type": "SHOW_POPUP_INFO",
        },
        Object {
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "OK",
              "onClick": [Function],
            },
          ],
          "callbacks": [MockFunction],
          "caption": "WARNING",
          "checkbox": Object {
            "callbacks": [MockFunction],
            "text": "Don't Show this next time.",
          },
          "checkboxCallbacks": [MockFunction],
          "checkboxText": "Don't Show this next time.",
          "message": "Some texts were changed to other Fonts when parsing texts to paths and some character may not converted normally.",
          "type": "SHOW_POPUP_WARNING",
        },
        Object {
          "buttonType": "RETRY_CANCEL",
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "Retry",
              "onClick": [Function],
            },
            Object {
              "className": "btn-default",
              "label": "Cancel",
              "onClick": [Function],
            },
          ],
          "caption": "UH-OH",
          "checkboxCallbacks": null,
          "checkboxText": null,
          "id": "monitor-reconnect",
          "message": "The connection with the machine has broken. Do you want to reconnect?",
          "onRetry": [MockFunction],
          "type": "SHOW_POPUP_ERROR",
        },
        Object {
          "buttonType": "CONFIRM_CANCEL",
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "Confirm",
              "onClick": [Function],
            },
            Object {
              "className": "btn-default",
              "label": "Cancel",
              "onClick": [Function],
            },
          ],
          "caption": undefined,
          "checkboxCallbacks": null,
          "checkboxText": null,
          "message": "This will load arrangement of presets and replacing customized parameters set in the file, are you sure to proceed?",
          "onConfirm": [MockFunction],
        },
        Object {
          "buttonLabels": Array [
            "UPDATE",
          ],
          "buttonType": "CUSTOM_CANCEL",
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "UPDATE",
              "onClick": [Function],
            },
            Object {
              "className": "btn-default",
              "label": "Cancel",
              "onClick": [Function],
            },
          ],
          "callbacks": [MockFunction],
          "caption": "Machine firmware Update",
          "checkboxCallbacks": null,
          "checkboxText": null,
          "id": "latest-firmware",
          "message": "You have the latest Machine firmware",
          "onCancel": [MockFunction],
        },
        Object {
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "Set Connection",
              "onClick": [MockFunction],
            },
            Object {
              "className": "btn-default primary",
              "label": "Retry",
              "onClick": [MockFunction],
            },
          ],
          "caption": undefined,
          "checkboxCallbacks": null,
          "checkboxText": null,
          "message": "Unable to find machine ",
        },
      ],
      "popById": [Function],
      "popFromStack": [Function],
    }
  }
>
  <Children />
</ContextProvider>
`;

exports[`should render correctly 8`] = `
<ContextProvider
  value={
    Object {
      "alertProgressStack": Array [
        Object {
          "caption": "prepared",
          "id": "check-status",
          "isProgress": true,
          "message": "prepared",
          "type": "NONSTOP",
        },
        Object {
          "buttonType": "YES_NO",
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "Yes",
              "onClick": [Function],
            },
            Object {
              "className": "btn-default",
              "label": "No",
              "onClick": [Function],
            },
          ],
          "caption": "INFO",
          "checkboxCallbacks": null,
          "checkboxText": null,
          "message": "File already exists, do you want to replace it?",
          "onYes": [MockFunction],
          "type": "SHOW_POPUP_INFO",
        },
        Object {
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "OK",
              "onClick": [Function],
            },
          ],
          "callbacks": [MockFunction],
          "caption": "WARNING",
          "checkbox": Object {
            "callbacks": [MockFunction],
            "text": "Don't Show this next time.",
          },
          "checkboxCallbacks": [MockFunction],
          "checkboxText": "Don't Show this next time.",
          "message": "Some texts were changed to other Fonts when parsing texts to paths and some character may not converted normally.",
          "type": "SHOW_POPUP_WARNING",
        },
        Object {
          "buttonType": "RETRY_CANCEL",
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "Retry",
              "onClick": [Function],
            },
            Object {
              "className": "btn-default",
              "label": "Cancel",
              "onClick": [Function],
            },
          ],
          "caption": "UH-OH",
          "checkboxCallbacks": null,
          "checkboxText": null,
          "id": "monitor-reconnect",
          "message": "The connection with the machine has broken. Do you want to reconnect?",
          "onRetry": [MockFunction],
          "type": "SHOW_POPUP_ERROR",
        },
        Object {
          "buttonType": "CONFIRM_CANCEL",
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "Confirm",
              "onClick": [Function],
            },
            Object {
              "className": "btn-default",
              "label": "Cancel",
              "onClick": [Function],
            },
          ],
          "caption": undefined,
          "checkboxCallbacks": null,
          "checkboxText": null,
          "message": "This will load arrangement of presets and replacing customized parameters set in the file, are you sure to proceed?",
          "onConfirm": [MockFunction],
        },
        Object {
          "buttonLabels": Array [
            "UPDATE",
          ],
          "buttonType": "CUSTOM_CANCEL",
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "UPDATE",
              "onClick": [Function],
            },
            Object {
              "className": "btn-default",
              "label": "Cancel",
              "onClick": [Function],
            },
          ],
          "callbacks": [MockFunction],
          "caption": "Machine firmware Update",
          "checkboxCallbacks": null,
          "checkboxText": null,
          "id": "latest-firmware",
          "message": "You have the latest Machine firmware",
          "onCancel": [MockFunction],
        },
        Object {
          "buttons": Array [
            Object {
              "className": "btn-default primary",
              "label": "Set Connection",
              "onClick": [MockFunction],
            },
            Object {
              "className": "btn-default primary",
              "label": "Retry",
              "onClick": [MockFunction],
            },
          ],
          "caption": undefined,
          "checkboxCallbacks": null,
          "checkboxText": null,
          "message": "Unable to find machine ",
        },
      ],
      "popById": [Function],
      "popFromStack": [Function],
    }
  }
>
  <Children />
</ContextProvider>
`;
